### PR TITLE
#89814 - introduce retry-sleep-durations in webhook configuration

### PR DESCRIPTION
**Pull request by mszymeshop**
This PR does not enable usage of the config yet in Http Sender

### CODE REVIEW

**Review by dawidkeshop — APPROVED** [review]

File: src/CaptainHook.EventHandlerActor/Handlers/SendRequest.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+using System;
+using System.Net.Http;
+using JetBrains.Annotations;
+
+namespace CaptainHook.EventHandlerActor.Handlers
+{
+    public class SendRequest
```
Comment (chomick) [inline src/CaptainHook.EventHandlerActor/Handlers/SendRequest.cs:7]: 👍

**Review by tabman83 — APPROVED** [review]

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs [hunk 2]
```diff
@@ -111,7 +111,7 @@ public async Task CheckWebhookCall()
             await webhookResponseHandler.CallAsync(messageData, new Dictionary<string, object>(), _cancellationToken);
 
             _mockAuthHandlerFactory.Verify(e => e.GetAsync(messageData.SubscriberConfig, _cancellationToken), Times.Once);
-            _mockHttpSender.Verify(x => x.SendAsync(HttpMethod.Post, new Uri(ExpectedWebHookUri), It.IsAny<WebHookHeaders>(), It.IsAny<string>(), It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()), Times.Once);
+            _mockHttpSender.Verify(x => x.SendAsync(It.Is<SendRequest>(request => request.Uri == new Uri(ExpectedWebHookUri)), It.IsAny<CancellationToken>()), Times.Once);
```
Comment (chomick) [inline src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs:114]: We should also verify the http verb as before?

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs [hunk 3]
```diff
@@ -36,28 +36,28 @@ public async Task<TokenResponse> RequestClientCredentialsTokenAsync(ClientCreden
             return await httpClient.RequestClientCredentialsTokenAsync(request, cancellationToken);
         }
 
-        public async Task<HttpResponseMessage> SendAsync(HttpMethod httpMethod, Uri uri, WebHookHeaders headers, string payload, TimeSpan timeout = default, CancellationToken cancellationToken = default)
+        public async Task<HttpResponseMessage> SendAsync(SendRequest sendRequest, CancellationToken cancellationToken = default)
```
Comment (chomick) [inline src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:39]: Any reason we are actually not using the intervals/retries here?
Should we also add unit tests to cover empty list (no retries) and non-empty one (if not already there)

**Review by nikhilgirraj — APPROVED** [review]

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs (hunk 3 above)
Comment (mszymeshop) [reply src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:39]: The Retries are part of the SendRequest dto. Will add more tests to cover the scenario you mentioned.

Comment (mszymeshop) [reply src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:39]: I understand the need for tests, this class however hasn't been tested so far. It is not the easiest to test either. Will definitely add complexity and time.

File: src/CaptainHook.Common/Configuration/EventHandlerConfig.cs [hunk 4]
```diff
@@ -83,6 +86,12 @@ public string HttpVerb
         [JsonProperty(Order = 6)]
         public TimeSpan Timeout { get; set; } = new TimeSpan(0, 0, 100);
 
+        /// <summary>
+        /// Retry sleep durations to use for request retry logic
+        /// </summary>
+        [JsonProperty(Order = 7)]
+        public TimeSpan[] RetrySleepDurations { get; set; }
```
[the remark this answers is not in the document]
Comment (mszymeshop) [reply src/CaptainHook.Common/Configuration/EventHandlerConfig.cs:93]: Here is the result of serialisation from API:
```js
"routes": [
  {
    "name": null,
    "uri": "http://blah.blah",
    "selector": "*",
    "authenticationConfig": {
      "username": "user1",
      "password": "***",
      "type": 1
    },
    "httpVerb": "POST",
    "webhookRequestRules": [],
    "timeout": "00:01:40",
    "retrySleepDurations": [
      "00:00:20",
      "00:00:30"
    ]
  }
]
```

File: src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs (hunk 2 above)
Comment (mszymeshop) [reply src/Tests/CaptainHook.Tests/Web/WebHooks/WebhookResponseHandlerTests.cs]: Missed it, added now.

File: src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs (hunk 3 above)
Comment (mszymeshop) [reply src/CaptainHook.EventHandlerActor/Handlers/HttpSender.cs:39]: I updated the name of the PR to clearly identify the purpose. The next PR will focus on the actual usage of the config.
